pkg/kubelet/cm/dra/plugin/plugin.go

```diff
@@ -18,94 +18,140 @@ package plugin
-		grpc.WithContextDialer(func(ctx context.Context, target string) (net.Conn, error) {
-			return (&net.Dialer{}).DialContext(ctx, network, target)
-		}),
-	)
-	if err != nil {
-		return nil, err
-	}
-
-	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
-	defer cancel()
-
-	if ok := conn.WaitForStateChange(ctx, connectivity.Connecting); !ok {
-		return nil, errors.New("timed out waiting for gRPC connection to be ready")
-	}
-
-	p.conn = conn
-	return p.conn, nil
-}
-
 // RegistrationHandler is the handler which is fed to the pluginwatcher API.
 type RegistrationHandler struct {
-	controller *nodeResourcesController
+	ctx        context.Context
+	kubeClient kubernetes.Interface
+	restConfig *rest.Config
+	getNode    func() (*v1.Node, error)
 }
 
+var _ cache.PluginHandler = &RegistrationHandler{}
+
 // NewPluginHandler returns new registration handler.
 //
 // Must only be called once per process because it manages global state.
 // If a kubeClient is provided, then it synchronizes ResourceSlices
 // with the resource information provided by plugins.
-func NewRegistrationHandler(kubeClient kubernetes.Interface, getNode func() (*v1.Node, error)) *RegistrationHandler {
-	handler := &RegistrationHandler{}
+func NewRegistrationHandler(kubeClient kubernetes.Interface, restConfig *rest.Config, getNode func() (*v1.Node, error)) *RegistrationHandler {
+	handler := &RegistrationHandler{
+		// The context and thus logger should come from the caller.
+		ctx:        klog.NewContext(context.TODO(), klog.LoggerWithName(klog.TODO(), "DRA registration handler")),
+		kubeClient: kubeClient,
+		restConfig: restConfig,
+		getNode:    getNode,
+	}
 
-	// If kubelet ever gets an API for stopping registration handlers, then
-	// that would need to be hooked up with stopping the controller.
-	handler.controller = startNodeResourcesController(context.TODO(), kubeClient, getNode)
+	// When kubelet starts up, no DRA driver has registered yet. None of
+	// the drivers are usable until they come back, which might not happen
+	// at all. Therefore it is better to not advertise any local resources
+	// because pods could get stuck on the node waiting for the driver
+	// to start up.
+	//
+	// This has to run in the background.
+	go handler.wipeResourceSlices("")
 
 	return handler
 }
 
+// wipeResourceSlices deletes ResourceSlices of the node, optionally just for a specific driver.
+func (h *RegistrationHandler) wipeResourceSlices(pluginName string) {
+	if h.restConfig == nil {
+		return
+	}
+	ctx := h.ctx
+	logger := klog.FromContext(ctx)
+	backoff := wait.Backoff{
+		Duration: time.Second,
+		Factor:   2,
+		Jitter:   0.2,
+		Cap:      5 * time.Minute,
+		Steps:    100,
+	}
+	discoveryCache := memory.NewMemCacheClient(h.kubeClient.Discovery())
+	restMapper := restmapper.NewDeferredDiscoveryRESTMapper(discoveryCache)
+	gk := resourceapi.SchemeGroupVersion.WithKind("ResourceSlice").GroupKind()
+	dynClient, err := dynamic.NewForConfig(h.restConfig)
```

Comment:
prefer constructing and persisting the dynamic client at the kubelet level rather than passing around a REST config

Reply:
actually, I'm not sure this needs a dynamic client any more... see below if the only call we're making is DeleteCollection